### PR TITLE
Widening range of grpcio versions allowed.

### LINKED ISSUE #27299: [Core] investigate why Ray hangs with grpcio==1.48.0
### What happened + What you expected to happen

We need to dive deep down to why `ray.init` hangs with grpcio==1.48.0.

Bonus point: we need to figure out our strategy moving forward, since grpcio has been caused several issues.
https://github.com/ray-project/ray/commit/502c3e132d27ce5106fe98ecde8deacbdde57e6a
https://github.com/ray-project/ray/commit/e3051ebf673a6bb36e0c380b121c08c1b5fd0f4c
https://github.com/ray-project/ray/commit/fda345335ae5fe4d8ff50927f5b2c6afa3e9b56d

### Versions / Dependencies

latest ray

### Reproduction script

pip install ray
pip install grpcio==1.48.0

python
```
import ray
ray.init()
```
This will hang.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by cadedaniel**
See https://github.com/ray-project/ray/issues/27299 for context.

### script.py
Testing with the following code:
```python3
#!/usr/bin/env python3
import ray
ray.init()


@ray.remote
def task(argument):
    import grpc
    import platform
    print(argument, grpc.__version__, platform.python_version())

ray.get(task.remote('Hello world'))
```
### python/grpcio versions tested on above script
```
grpcio==1.43.0 python==3.10.4
grpcio==1.44.0 python==3.10.4
grpcio==1.45.0 python==3.10.4
grpcio==1.46.0 python==3.10.4
grpcio==1.47.0 python==3.10.4
grpcio==1.48.1 python==3.10.4
grpcio==1.43.0 python==3.6.13
grpcio==1.44.0 python==3.6.13
grpcio==1.45.0 python==3.6.13
grpcio==1.46.0 python==3.6.13
grpcio==1.47.0 python==3.6.13
grpcio==1.48.1 python==3.6.13
grpcio==1.43.0 python==3.7.13
grpcio==1.44.0 python==3.7.13
grpcio==1.45.0 python==3.7.13
grpcio==1.46.0 python==3.7.13
grpcio==1.47.0 python==3.7.13
grpcio==1.48.1 python==3.7.13
grpcio==1.43.0 python==3.8.13
grpcio==1.44.0 python==3.8.13
grpcio==1.45.0 python==3.8.13
grpcio==1.46.0 python==3.8.13
grpcio==1.47.0 python==3.8.13
grpcio==1.48.1 python==3.8.13
grpcio==1.43.0 python==3.9.13
grpcio==1.44.0 python==3.9.13
grpcio==1.45.0 python==3.9.13
grpcio==1.46.0 python==3.9.13
grpcio==1.47.0 python==3.9.13
grpcio==1.48.1 python==3.9.13
```

### test code

```bash
#!/usr/bin/env bash

set -e

source /home/ray/anaconda3/etc/profile.d/conda.sh

pr_commit="0193a19226c29c9988760114d67f6ea9af99f9e7"

ray_wheels="$(aws s3 ls s3://ray-ci-artifact-pr-public/$pr_commit/tmp/artifacts/.whl/ | grep -v 'cpp' | awk '{print $4}')"
grpcio_versions="1.43 1.44 1.45 1.46 1.47 1.48.1"

for ray_wheel in $ray_wheels; do

    conda_create_cmd=$(echo $ray_wheel | sed 's/-/ /'g | awk '{print $3}' | sed 's/cp//g' | sed 's/3/3\./g' | sed 's/^/conda create -n temp python=/g')
    $conda_create_cmd --yes
    conda activate temp
    
    for grpcio_version in $grpcio_versions; do
        printf "Uninstalling\n"
        pip uninstall grpcio -y
        pip uninstall ray -y

        printf "Installing grpcio_version $grpcio_version\n"
        pip install grpcio==$grpcio_version

        printf "Installing Ray wheel $ray_wheel"
        pip install "https://ray-ci-artifact-pr-public.s3.us-west-2.amazonaws.com/$pr_commit/tmp/artifacts/.whl/$ray_wheel"

        printf "Running script on grpcio_version $grpcio_version\n"
        ./script.py
    done
done
```

### CODE REVIEW

**Comment by cadedaniel** [comment]
@jjyao what is the best way to test the `setup.py` installation process? I need to test python<3.10 and python>=3.10, is conda envs the best way?

File: python/requirements.txt [hunk 1]
```diff
@@ -12,8 +12,8 @@ cloudpickle
 filelock
 frozenlist
 gpustat == 1.0.0b1
-grpcio >= 1.32.0, <= 1.43.0; python_version < '3.10'
-grpcio >= 1.42.0, <= 1.43.0; python_version >= '3.10'
+grpcio >= 1.32.0, <= 1.48.1; python_version < '3.10'
```
Comment (jjyao) [inline python/requirements.txt:15]: Why can't we go above 1.48.1?

**Comment by jjyao** [comment]
> what is the best way to test the setup.py installation process? I need to test python<3.10 and python>=3.10, is conda envs the best way?

Yea, I would use conda for that.

File: python/requirements.txt (hunk 1 above)
Comment (cadedaniel) [reply python/requirements.txt:15]: This is the highest version currently published. Do you think I should remove the upper bound altogether?

Comment (jjyao) [reply python/requirements.txt:15]: Yes, we should remove it so future higher versions can be supported automatically.

**Review by jjyao — APPROVED** [review]

**Review by richardliaw — APPROVED** [review]
stamp

**Comment by jjyao** [comment]
Could you pull the master head? Want to make sure failed tests are unrelated.

**Review by scv119 — CHANGES_REQUESTED** [review]
hmm, we should still set the upper bound? for core dependencies, we should only approve a version until it's proven innocent.

**Comment by cadedaniel** [comment]
> hmm, we should still set the upper bound? for core dependencies, we should only approve a version until it's proven innocent.

cc @jjyao

**Comment by jjyao** [comment]
> hmm, we should still set the upper bound? for core dependencies, we should only approve a version until it's proven innocent.

@scv119 This is not the strategy we are following due to the nature of Ray as a library (if you look at setup.py, we don't have upper bounds for most of (core) dependencies). Ray shouldn't prevent users from using the latest versions of dependencies (they may want to use that for some reason like bug fix) that might come out after Ray's release.

**Comment by scv119** [comment]
hmm this has been the strategy we have been following for grpcio right? Empirically, for grpcio, not setting upper bound has caused a lot of problems, comparing to the issues where users are not able to upgrade to the latest grpcio.

**Comment by jjyao** [comment]
grpcio doesn't have upper bound by default as well. I think what we did in the past was forcing an upper bound when we discover an issue and lift the upper bound afterwards after the issue is fixed. (this is what we normally do for other core dependencies as well).

The strategy I mentioned is a general strategy we apply to all of our dependencies. Whether we want to make an exception for grpcio is a separate story but I don't think it should apply to all dependencies.

Personally I still feel we shouldn't put an upper bound (we may put an upper bound on the major version number not but the minor version number if they are following semantic versioning).

@richardliaw may have more insights on this.

**Comment by cadedaniel** [comment]
Sounds good.

There are more non-flaky Windows tests failing, taking note here so see if they repeat:
```
Windows [1/6] (first try):
* //python/ray/tests:test_multiprocessing (timeout)
Windows [1/6] (second try):
* //python/ray/tests:test_multiprocessing  (timeout)
* //python/ray/tests:test_asyncio  (timeout)

Windows [5/6] (first try):
* //python/ray/tests:test_queue (timeout)

Windows [5/6] (second try):
* //python/ray/tests:test_queue (timeout)
* //python/ray/tests:test_runtime_env_working_dir_3 (failure for something to be GC'd after 20 seconds)
```

**Comment by cadedaniel** [comment]
`//python/ray/tests:test_multiprocessing test_callbacks` hangs for both grpcio==1.43.0 and 1.48.1 so I assume it has to do with my environment?

Is there a way I can replicate the BK environment and iterate faster? It is taking forever to wait for BK

**Comment by h-vetinari** [comment]
Thanks for working on this! Just to understand - besides the missing timeout in the tests, there were no other changes necessary to support grpc 1.48?

**Comment by cadedaniel** [comment]
> Thanks for working on this! Just to understand - besides the missing timeout in the tests, there were no other changes necessary to support grpc 1.48?

Yep! I believe we prohibited this version because it was causing the hang; since 1.48.0 was yanked and 1.48.1 fixes the issue, we can simply allow it.

**Comment by cadedaniel** [comment]
I'm increasing the timeouts on all of the tests that I've seen become flaky on Windows because of this change.
```
test_reference_counting
test_multiprocessing
test_asyncio
test_queue
```

**Review by scv119 — APPROVED** [review]

**Comment by cadedaniel** [comment]
I have a lint fix but waiting for windows builds to succeed before I push

**Comment by jjyao** [comment]
windows tests passed

**Pull request merged**